include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h

```diff
@@ -32,6 +32,117 @@ namespace LinearAlgebra
 
   namespace TpetraWrappers
   {
+    namespace internal
+    {
+      template <typename Number, typename MemorySpace>
+      void
+      apply(const SparseMatrix<Number, MemorySpace> &M,
+            const Vector<Number, MemorySpace>       &src,
+            Vector<Number, MemorySpace>             &dst,
+            Teuchos::ETransp                         mode = Teuchos::NO_TRANS,
+            Number alpha = Teuchos::ScalarTraits<Number>::one(),
+            Number beta  = Teuchos::ScalarTraits<Number>::zero())
+      {
+        Assert(&src != &dst,
+               SparseMatrix<double>::ExcSourceEqualsDestination());
+        Assert(M.trilinos_matrix().isFillComplete(),
+               SparseMatrix<double>::ExcMatrixNotCompressed());
+
+        Assert(src.trilinos_vector().getMap()->isSameAs(
+                 *M.trilinos_matrix().getDomainMap()),
+               SparseMatrix<double>::ExcColMapMissmatch());
+        Assert(dst.trilinos_vector().getMap()->isSameAs(
+                 *M.trilinos_matrix().getRangeMap()),
+               SparseMatrix<double>::ExcDomainMapMissmatch());
+
+        M.trilinos_matrix().apply(
+          src.trilinos_vector(), dst.trilinos_vector(), mode, alpha, beta);
+      }
+
+
+
+      template <typename Number, typename MemorySpace>
+      void
+      apply(const SparseMatrix<Number, MemorySpace> &M,
+            const dealii::Vector<Number>            &src,
+            dealii::Vector<Number>                  &dst,
+            Teuchos::ETransp                         mode = Teuchos::NO_TRANS,
+            Number alpha = Teuchos::ScalarTraits<Number>::one(),
+            Number beta  = Teuchos::ScalarTraits<Number>::zero())
+      {
+        Assert(&src != &dst,
+               SparseMatrix<double>::ExcSourceEqualsDestination());
+        Assert(M.trilinos_matrix().isFillComplete(),
+               SparseMatrix<double>::ExcMatrixNotCompressed());
+
+#  ifdef DEBUG
+        // TODO: If the host spaces do not match, we can copy the memory
+        //       from one device to an other.
+        bool memory_space_matches =
+          std::is_same_v<typename MemorySpace::kokkos_space,
+                         typename Kokkos::HostSpace>;
+        Assert(memory_space_matches, ExcInternalError());
+#  endif
```

Comment:
We can just do this properly.

Reply:
I applied the two patches below as you suggested and removed this assert because it is no longer necessary.